packages/discount-code-exporter/src/main.js

```diff
@@ -1,3 +1,182 @@
-  return a + b
+/* @flow */
+import type {
+  ApiConfigOptions,
+  CodeData,
+  ImporterOptions,
+  LoggerOptions,
+} from 'types/discountCodes'
+import type {
+  Client,
+  ClientRequest,
+} from 'types/sdk'
+import { createClient } from '@commercetools/sdk-client'
+import { createRequestBuilder } from '@commercetools/api-request-builder'
+import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
+import {
+  createAuthMiddlewareForClientCredentialsFlow,
+} from '@commercetools/sdk-middleware-auth'
+import {
+  createUserAgentMiddleware,
+} from '@commercetools/sdk-middleware-user-agent'
+import csv from 'fast-csv'
+import JSONStream from 'JSONStream'
+import { flatten } from 'flat'
+import pkg from '../package.json'
+
+type ConfigType = {
+  batchSize: number;
+  accessToken: string;
+  delimiter: string;
+  exportFormat: string;
+  predicate: string;
+  multiValueDelimiter: string
+}
+export default class DiscountCodeExport {
+  // Set flowtype annotations
+  apiConfig: ApiConfigOptions;
+  client: Client;
+  config: ConfigType;
+  logger: LoggerOptions;
+  _processCode: Function;
+
+  constructor (
+    options: ImporterOptions,
+    logger: LoggerOptions,
+  ) {
+    if (!options.apiConfig)
+      throw new Error('The contructor must be passed an `apiConfig` object')
+    if (options.batchSize > 500)
+      throw new Error('The `batchSize` must not be more than 500')
+    this.apiConfig = options.apiConfig
+    this.client = createClient({
+      middlewares: [
+        createAuthMiddlewareForClientCredentialsFlow(this.apiConfig),
+        createUserAgentMiddleware({
+          libraryName: pkg.name,
+          libraryVersion: pkg.version,
+        }),
+        createHttpMiddleware({ host: this.apiConfig.apiUrl }),
+      ],
+    })
+
+    const defaultOptions = {
+      delimiter: ',',
+      multiValueDelimiter: ';',
+      batchSize: 500,
+      exportFormat: 'json',
+    }
+
+    this.config = { ...defaultOptions, ...options }
+
+    this.logger = {
+      error: () => {},
+      warn: () => {},
+      info: () => {},
+      verbose: () => {},
+      ...logger,
+    }
+
+    this._processCode = this._processCode.bind(this)
+  }
+
+  run (outputStream: stream$Writable) {
+    this.logger.info('Starting Export')
+    if (this.config.exportFormat === 'csv') {
+      const csvOptions = {
+        headers: true,
+        delimiter: this.config.delimiter,
+      }
+      const csvStream = csv
+        .createWriteStream(csvOptions)
+        .transform(this._processCode)
+      csvStream.pipe(outputStream)
+      this._handleOutput(outputStream, csvStream)
+    } else {
+      const jsonStream = JSONStream.stringify()
+      jsonStream.pipe(outputStream)
+      this._handleOutput(outputStream, jsonStream)
+    }
+  }
+
+  _handleOutput (
+    outputStream: stream$Writable,
+    pipeStream: stream$Writable,
+  ) {
+    this._fetchCodes(pipeStream)
+      .then(() => {
+        this.logger.info('Operation completed successfully')
```

Comment:
Can you be more descriptive in a log message? So we will know what `operation` completed successfully?